File: arrays/56. Merge Intervals/merge_intervals.py
class Solution:
    """Leetcode https://leetcode.com/problems/merge-intervals/ problem's solution"""

    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        result = list()
        # sort the given list, sorted times are easy to calculate
        intervals.sort()

        for elements in intervals:
            # first element condition
            if len(result) < 1:
                result.append(elements)
            else:
                # take the previous time slot for comparision
                prev_element = result[-1]

                # if the current time slot can be included in previous time slot
                # then take the max
                if prev_element[1] >= elements[0]:
                    prev_element[1] = max(prev_element[1], elements[1])
                else:
                    # found new time slot
                    result.append(elements)

        return result

File: arrays/56. Merge Intervals/merge_intervals.rb
def merge(intervals)
 merged = []
  # sort intervals
  intervals.sort_by! { |interval| interval[0] }

  intervals.each do |interval|
    if merged.empty? || merged.last[1] < interval[0]
      merged << interval
    else
      merged.last[1] = interval[1] if interval[1] > merged.last[1]
    end
  end
  merged
end

# merge([[1,3],[2,6],[8,10],[15,18]])